Fix/delayed sto

martiniil opened the pull request:
fix #151

martiniil committed:
Call recover before unhold

* adjust StoModbusAdapter
* change expected order in unit and integration test

martiniil committed:
Add STOExecutor

martiniil committed:
Add STOExecutor unit test

martiniil committed:
Add ServiceClientFactory

martiniil committed:
Add ServiceClientMock

jschleicher commented on prbt_hardware_support/include/prbt_hardware_support/service_client_factory.h:
```diff
@@ -0,0 +1,48 @@
+  template <typename Service>
+  inline static ros::ServiceClient create(const std::string &name)
+  {
+    waitForService(name);
```
better add `while(ros::ok) { ROS_INFO_THROTTLED...}` as in http://wiki.ros.org/roscpp/Overview/Logging to give feedback to the user while waiting.
see https://github.com/PilzDE/pilz_robots/blob/10b4fb75ee28a35107087a024a174cd67026305f/prbt_hardware_support/include/prbt_hardware_support/wait_for_service.h#L38-L43

martiniil replied:
this is exactly the waitForService() that is called ...

jschleicher replied:
okay. To avoid further confusion I'd suggest to unify both. Either keep your Factory class and refactor  braketest accordingly or stick to Hagen's functional approach and remove your class. Why would you need two approaches (files, tests, documentation) for the same functionality?
https://github.com/PilzDE/pilz_robots/blob/8bd4e2042462b7dc2768748bded6e2e150d234cb/prbt_hardware_support/src/brake_test_executor.cpp#L49-L58

martiniil replied:
I agree. I would suggest to do the refactoring in a separate PR and concentrate on the STO fix here.

martiniil replied:
Also I think the confusion is between ros::service::waitForService() and prbt_hardware_support::waitForService(), which can be eliminated by refactoring.

jschleicher replied:
Is there already a PR for `waitForService` refactoring available? Or an open issue?

jschleicher replied:
What about moving service_client_factory into a central package for easy re-usability?

jschleicher reviewed (commented):
Where is your state diagram gone? Didn't you have an image in a previous version of this branch?

martiniil committed:
Add state diagram

martiniil committed:
plan unittest sto executor

martiniil committed:
Complete unittest of sto exeutor

martiniil committed:
Refactor STOExecutor. passes unit test.

martiniil committed:
Fix formatting

martiniil committed:
Merge branch 'melodic-devel' into fix/delayed_sto

martiniil committed:
Refactor modbus sto adapter

martiniil committed:
Increase line and function coverage

martiniil committed:
Changes to spec

martiniil committed:
Add newline at end of file

martiniil committed:
Minor fix

martiniil committed:
Changes to spec

* Describe exactly what we mean by stop1 in the spec
* Add special case "STO false during recover"

martiniil committed:
Replace robot -> manipulator in spec

agutenkunst commented on pilz_testutils/include/pilz_testutils/async_test.h:
```diff
@@ -88,6 +88,11 @@ class AsyncTest
      */
     void barricade(std::initializer_list<std::string> clear_events);
 
+    /**
```
Not used please drop.

agutenkunst commented on pilz_testutils/src/async_test.cpp:
```diff
@@ -52,4 +55,23 @@ void AsyncTest::triggerClearEvent(std::string event)
   cv_.notify_one();
 }
 
+void AsyncTest::printAsyncEvents()
```
As long not used drop it please

jschleicher commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -100,9 +143,201 @@ class AdapterSto
+  {
+    if (next_action_ == Action::HALTING)
+    {
+      next_action_ = Action::RECOVERING;
```
The transition HALTING->RECOVERING isn't supported.
Please also correct diag_state_sto_executor_action.png

martiniil replied:
What do you mean? This triggers recover when STO false -> true.

jschleicher replied:
Exactly. But the shortcut is only desired for RECOVERING->HALTING. Once STO goes to false, the full halt is always executed (see doors specification of PSS4000 function block).

martiniil replied:
Sorry, if the code is misleading, I am rewriting this part at the moment and hope to resolve the ambiguities.

The full halt (when STO goes false) is UNHOLDING -> HOLDING -> HALTING

HALTING -> RECOVERING for STO true is no shortcut

jschleicher replied:
Or: Do you deliberately deviate from the PSS4000 spec and leave that "full stop" to the safety controller? As said: In my opinion the diag_state_sto_executor_action.png should be asymmetric: Only having one shortcut for RECOVERING->HALTING. In state HOLDING, the sto=true could be visualized as self-transition / effectively a no-op.

martiniil replied:
Ok, I got it. Thanks!

jschleicher commented on prbt_hardware_support/test/include/prbt_hardware_support/service_client_mock.h:
```diff
@@ -0,0 +1,92 @@
+ * @tparam S service type
+ */
+template <typename S>
+class ServiceClientMock
```
Is the ServiceClientMock class useful in more general and could be placed inside test_utils?

martiniil replied:
yes!

jschleicher reviewed (commented):
Thanks @martiniil for pushing this. I didn't look at the unittest. Just two inline comments:

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -55,43 +86,55 @@ class AdapterSto
   void updateSto(const bool sto);
 
 private:
+  enum Action
```
`enum class` maybe?

martiniil replied:
outdated.

martiniil committed:
Update hold requirements in spec

also explain hold (no requirement, since it is not tested in prbt_hardware_support)

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -100,9 +143,201 @@ class AdapterSto
+  ros::Rate rate(10);
+  while (!terminate_)
+  {
+    {
```
This part is hard to understand. As far as I understand you only need to lock the assignment.

martiniil replied:
outdated.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -100,9 +143,201 @@ class AdapterSto
+  while (!terminate_)
+  {
+    {
+      std::lock_guard<std::mutex> lock(action_mutex_);
```
Are you really sure you need this mutex. The enum should be atomic and you have on thread writting it right?

martiniil replied:
outdated.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -100,9 +143,201 @@ class AdapterSto
+  run_thread_ = std::thread{[this] { this->run(); }};
+}
+
+template <class T>
```
Please provide a detailed description here

martiniil replied:
done.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -40,12 +42,41 @@ namespace prbt_hardware_support
  */
-class AdapterSto
+template <class T = ros::ServiceClient>
+class AdapterStoTemplated
```
Please add move detail to the class description on what you are trying to achieve

martiniil replied:
done.

agutenkunst commented on prbt_hardware_support/doc/diag_state_sto_executor_action.uxf:
```diff
@@ -0,0 +1,195 @@
+<?xml version="1.0" encoding="UTF-8"?><diagram program="umlet" version="13.3">
```
In your diagram as far as I understand some guard conditions could provide more insight?

martiniil replied:
outdated.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -100,9 +143,201 @@ class AdapterSto
+{
+  std::lock_guard<std::mutex> lock(action_mutex_);
+
+  if (next_action_ == finished_action)
```
Not readable.

martiniil replied:
outdated.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -100,9 +143,201 @@ class AdapterSto
+
+  if (next_action_ == finished_action)
+  {
+    switch (finished_action)
```
Maybe some StateMachine Framework would help keeping this sane?

martiniil replied:
yes!

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -100,9 +143,201 @@ class AdapterSto
+{
+  bool success{false};
+
+  switch (action)
```
Any idea on some pattern not relying on switch cases? As soon as some other ACTION is introduced these will grow and spread...

agutenkunst replied:
https://stackoverflow.com/a/126475

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_modbus_adapter_sto.cpp:
```diff
@@ -211,33 +212,36 @@ TEST_F(ModbusAdapterStoTest, testSetupNoDisableService)
   manipulator_.advertiseRecoverService(nh_, RECOVER_SERVICE_T);
 
-  modbus_sto_adapter_.reset(new ModbusAdapterSto(nh_, test_api_spec));
+  auto t = asyncConstructor();
```
variabel naming! Clean Code.

martiniil replied:
done.

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_modbus_adapter_sto.cpp:
```diff
@@ -249,9 +253,12 @@ TEST_F(ModbusAdapterStoTest, testSetupNoRecoverService)
   manipulator_.advertiseHaltService(nh_, HALT_SERVICE_T);
 
-  modbus_sto_adapter_.reset(new ModbusAdapterSto(nh_, test_api_spec));
+  auto t = asyncConstructor();
```
See above

martiniil replied:
Done

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_service_client_factory.cpp:
```diff
@@ -0,0 +1,67 @@
+ */
+TEST(ServiceClientFactoryIntegrationTest, testCreate)
+{
+  ros::AsyncSpinner spinner {1};
```
Are you sure you need this?

martiniil replied:
I think it is not needed.

martiniil committed:
Update state diagram

No shortcut when in holding: always halt & recover

martiniil committed:
Rename diagram files

martiniil committed:
WIP: Refactor adapter_sto

martiniil committed:
Add is_executing srv call to sto adapter

martiniil committed:
Update spec

martiniil committed:
Delete state diagram

martiniil committed:
Update architecture overview

martiniil committed:
WIP: add sequence diagram

agutenkunst commented:
I checked with b5bce3c5f2f4ccf3806233d1234bd7d96b552dd7

```
[DEBUG] [WallTime: 1563793840.023446066 /pilz_robots/prbt_hardware_support/src/pilz_modbus_client.cpp:161]: Sending new ROS-message.
Warning: recursive print statement has occurred.  Throwing out recursive print.
[ INFO] [WallTime: 1563793840.028309680 /ros_canopen/canopen_chain_node/src/ros_chain.cpp:132]: Recovering XXX
EMCY: 83#0000000000000000
EMCY: 84#0000000000000000
EMCY: 85#0000000000000000
EMCY: 86#0000000000000000
EMCY: 87#0000000000000000
EMCY: 88#0000000000000000
[DEBUG] [WallTime: 1563793840.267283104 /pilz_robots/prbt_hardware_support/src/pilz_modbus_client.cpp:161]: Sending new ROS-message.
Warning: recursive print statement has occurred.  Throwing out recursive print.
[ INFO] [WallTime: 1563793840.267619733 /pilz_robots/prbt_hardware_support/src/adapter_sto.cpp:68]: Performing STOP due to STO signal
[ INFO] [WallTime: 1563793840.470931269 /ros_canopen/canopen_chain_node/src/ros_chain.cpp:196]: Halting down XXX
```
so the time here would be ~45ms

I did a rebase against https://github.com/PilzDE/pilz_robots/pull/184 and got
```
[DEBUG] [WallTime: 1563797134.323560577 /pilz_robots/prbt_hardware_support/src/pilz_modbus_client.cpp:161]: Sending new ROS-message.
Warning: recursive print statement has occurred.  Throwing out recursive print.
[ERROR] [WallTime: 1563797134.323835277 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:166]: UPDATING IN THREAD 139794322971840
[ERROR] [WallTime: 1563797134.323937553 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:196]: Calling Recover (Service: /prbt/driver/recover)
[ INFO] [WallTime: 1563797134.325399648 /ros_canopen/canopen_chain_node/src/ros_chain.cpp:132]: Recovering XXX
[DEBUG] [WallTime: 1563797134.383626879 /pilz_robots/prbt_hardware_support/src/pilz_modbus_client.cpp:161]: Sending new ROS-message.
Warning: recursive print statement has occurred.  Throwing out recursive print.
[ERROR] [WallTime: 1563797134.383920953 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:166]: UPDATING IN THREAD 139794322971840
[ERROR] [WallTime: 1563797134.383966795 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:229]: Calling Hold on controller (Service: /prbt/manipulator_joint_trajectory_controller/hold)
[ WARN] [WallTime: 1563797134.385457115 /pilz_robots/pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h:99]: Trying to hold controller which is in hold mode already!
[ERROR] [WallTime: 1563797134.386955541 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:252]: Calling Halt on driver (Service: /prbt/driver/halt)
[ INFO] [WallTime: 1563797134.388161185 /ros_canopen/canopen_chain_node/src/ros_chain.cpp:196]: Halting down XXX
EMCY: 83#0000000000000000
EMCY: 84#0000000000000000
EMCY: 85#0000000000000000
EMCY: 86#0000000000000000
EMCY: 87#0000000000000000
EMCY: 88#0000000000000000
[ERROR] [WallTime: 1563797134.409911074 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:198]: Finished Recover (Service: /prbt/driver/recover)
[ WARN] [WallTime: 1563797134.409966943 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:208]: Lost STO Clearance during recover. Keeping controller in hold mode!
```

which should be what we want. But I can also produce a
```
[DEBUG] [WallTime: 1563797412.838820902 /pilz_robots/prbt_hardware_support/src/pilz_modbus_client.cpp:161]: Sending new ROS-message.
Warning: recursive print statement has occurred.  Throwing out recursive print.
[ERROR] [WallTime: 1563797412.839130300 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:166]: UPDATING IN THREAD 139743469784256
[ERROR] [WallTime: 1563797412.839230633 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:196]: Calling Recover (Service: /prbt/driver/recover)
[ INFO] [WallTime: 1563797412.840668482 /ros_canopen/canopen_chain_node/src/ros_chain.cpp:132]: Recovering XXX
EMCY: 83#0000000000000000
EMCY: 84#0000000000000000
EMCY: 85#0000000000000000
EMCY: 86#0000000000000000
EMCY: 87#0000000000000000
[DEBUG] [WallTime: 1563797412.926921287 /pilz_robots/prbt_hardware_support/src/pilz_modbus_client.cpp:161]: Sending new ROS-message.
Warning: recursive print statement has occurred.  Throwing out recursive print.
[ERROR] [WallTime: 1563797412.927224960 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:166]: UPDATING IN THREAD 139743469784256
[ERROR] [WallTime: 1563797412.927273668 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:229]: Calling Hold on controller (Service: /prbt/manipulator_joint_trajectory_controller/hold)
[ WARN] [WallTime: 1563797412.928813877 /pilz_robots/pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h:99]: Trying to hold controller which is in hold mode already!
EMCY: 88#0000000000000000
[ERROR] [WallTime: 1563797412.929511494 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:198]: Finished Recover (Service: /prbt/driver/recover)
[ WARN] [WallTime: 1563797412.929553917 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:208]: Lost STO Clearance during recover. Keeping controller in hold mode!
[ WARN] [WallTime: 1563797412.930461204 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:246]: Waiting before halt since trajectory is executing!
[ WARN] [WallTime: 1563797413.130625230 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:248]: Waiting done.
[ERROR] [WallTime: 1563797413.130678978 /pilz_robots/prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:252]: Calling Halt on driver (Service: /prbt/driver/halt)
[ INFO] [WallTime: 1563797413.132201856 /ros_canopen/canopen_chain_node/src/ros_chain.cpp:196]: Halting down XXX
```
which seems to indicate that we expect the controller to be executing introducing a 20ms extra wait. Not sure where this is coming from...

agutenkunst committed:
Add message on STO change during recover.

agutenkunst committed:
Merge branch 'melodic-devel' into fix/delayed_sto

agutenkunst committed:
Add debug output to ManipulatorMock

agutenkunst committed:
Enable debug output for ManipulatorMock

agutenkunst committed:
Add debug output. Set loglevel.

agutenkunst committed:
Set ROSCONSOLE_FORMAT

agutenkunst committed:
Add debug to async test

agutenkunst committed:
Remove unwanted debug output

agutenkunst committed:
WIP: Increase logging LEVEL of pilz_testutils

martiniil committed:
Add new state diagram

martiniil committed:
WIP: Make adapter a state machine

martiniil committed:
Minor update of state diagram

martiniil committed:
WIP: adapterSto as state machine

martiniil committed:
WIP. adapter_sto = state machine

martiniil committed:
wip: state machine + task queue

martiniil committed:
WIP: adapter_sto + state machine + task queue

martiniil committed:
Move service_client_mock to pilz_testutils

martiniil committed:
Remove unused function from async_test

martiniil committed:
Update documentation

martiniil committed:
Incorporate review comments

martiniil committed:
Fix waiting for hold traj execution

martiniil committed:
Remove unnecessary spinner

martiniil committed:
Merge branch 'melodic-devel' into fix/delayed_sto

martiniil committed:
Fix expectations in the case of failing service calls

martiniil committed:
Explain behaviour in case a service call fails

martiniil committed:
Fix unittest

martiniil committed:
Fix state machine

Since we want to react on every change of the sto-value, the state machine needs an aditional state StopPlusEnableRequestedDuringEnable.
However we could also argue that this is not necessary and no longer react on changes during stop.

martiniil committed:
Do not react on sto changes during stop.

After all this is not necessary, since a recurrent sto update is expected,
which takes effect once the stopping finished.
This simplifies the state machine significantly.

martiniil committed:
Add test for full line coverage

martiniil committed:
Add more tests for function full coverage

martiniil committed:
Delete outdated diagrams

martiniil committed:
Add test for full function coverage

martiniil committed:
Stabilize tests using the sto adapter

Do not expect that the adapter reacts on a single sto=true message when performing a stop,
but instead repeatedly send sto=true messages until the expected services have been called.

martiniil committed:
Fix clang-tidy error

martiniil committed:
clean up

martiniil commented:
I ran tests on both robots and compared b2a5c43 with 4ba39238983e42cc630628247cace5d9549a3381 (melodic-devel) 
The crucial part is the time between

> Sending new ROS-Message.

and 

> Halting down XXX

The outcome is as expected: Old (without fix): ~269ms, new: ~54ms

Old:
![melodic_devel](https://user-images.githubusercontent.com/35950815/62540080-06f69680-b857-11e9-8fb3-8370e5c238fe.png)

New:
![fix_delayed_sto](https://user-images.githubusercontent.com/35950815/62540090-0d850e00-b857-11e9-9b4e-2941fc20bfe1.png)

jschleicher commented on prbt_hardware_support/test/integrationtests/integrationtest_stop1.test:
```diff
@@ -45,6 +46,7 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
   <param name="halt_service_name" value="$(arg halt_service_name)"/>
   <param name="hold_service_name" value="$(arg hold_service_name)"/>
   <param name="unhold_service_name" value="$(arg unhold_service_name)"/>
+  <param name="is_executing_service_name" value="$(arg is_executing_service_name)"/>
```
Why do you prefer parameters for specifying service names over [remapping](http://wiki.ros.org/roslaunch/XML/remap)?

martiniil replied:
See #208

jschleicher commented on prbt_hardware_support/include/prbt_hardware_support/sto_state_machine.h:
```diff
@@ -0,0 +1,342 @@
+  // +---------------------------+--------------+---------------------------+--------------+----------+
+  Row< RobotInactive             , sto_updated  , Enabling                  , recover_start, sto_true >,
+  Row< RobotInactive             , sto_updated  , none                      , none         , sto_false>,
+  Row< Enabling                  , sto_updated  , none                      , none         , sto_true >,
```
Do you need those empty self-transitions with no action?

martiniil replied:
Yes, otherwise no_transition is called.
https://www.boost.org/doc/libs/1_64_0/libs/msm/doc/HTML/ch03s02.html#d0e471

jschleicher commented on prbt_hardware_support/test/include/prbt_hardware_support/pilz_manipulator_mock.h:
```diff
@@ -39,70 +39,131 @@ class ManipulatorMock
     void advertiseUnholdService(ros::NodeHandle nh, std::string unhold_service_name);
     void advertiseHaltService(ros::NodeHandle nh, std::string halt_service_name);
     void advertiseRecoverService(ros::NodeHandle nh, std::string recover_service_name);
+    void advertiseIsExecutingService(ros::NodeHandle nh, std::string is_executing_service_name);
```
Imho a better name would be `PilzControllerMock` since it mocks the ros_control-derived `PilzJointTrajectoryController`

martiniil replied:
It also mocks the canopen_chain_node.

jschleicher commented on prbt_hardware_support/test/integrationtests/integrationtest_modbus_adapter_sto.cpp:
```diff
@@ -318,46 +340,125 @@ TEST_F(ModbusAdapterStoTest, testRemoveUnholdService)
- * @tests{No_new_commands_during_STO_false,
- *  Tests that a message giving sto request leads to controller hold call.
+ * @tests{Stop1_Trigger,
+ *  Test that Stop 1 is triggered if STO value changes to false.
```
Seems like a Duplication of the unittest `TEST_F(AdapterStoTest, testEnableStopEnable)` and Integrationtest `TEST_F(Stop1IntegrationTest, testServiceCallbacks)`?

martiniil replied:
These are tests on different levels.

jschleicher reviewed (dismissed):
The arrows in prbt_hardware_support/doc/architecture_overview.png are overlapping:
![grafik](https://user-images.githubusercontent.com/29709205/62546231-65c20d00-b863-11e9-9f81-876e24ac5046.png)
Looks good overall, please see some inline comments (maybe resolve them into refactoring-issues).

martiniil committed:
Use persistent service for detecting hold traj execution

The frequent calls of the is_executing service introduced an overhead, which is reduced by making the connection persistent.

* Refactor AdapterSto
* Extend ServiceClientFactory for persistent service client connections
* Extend ServiceClientMock

martiniil committed:
Update architecture overview diagram

martiniil commented:
Related issue: https://github.com/PilzDE/pilz_robots/issues/206

martiniil committed:
Spin once for publish to take effect

agutenkunst committed:
Minor visual change in architecture diagram

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -15,94 +15,386 @@
+  /**
+   * @brief Call the service triggering recover of the drives.
+   */
+  void call_recover();
```
Please compare comment and function name. How about `callRecoverService`

martiniil replied:
Done

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -15,94 +15,386 @@
+  /**
+   * @brief Call the service triggering halt of the drives.
+   */
+  void call_halt();
```
See above

martiniil replied:
Done

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -15,94 +15,386 @@
+  /**
+   * @brief Call the service triggering hold of the controller. Wait until execution of the hold trajectory finished.
+   */
+  void call_hold();
```
See above

martiniil replied:
Done

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -15,94 +15,386 @@
+  /**
+   * @brief Call the service triggering unhold of the controller.
+   */
+  void call_unhold();
```
See above

martiniil replied:
Done

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -15,94 +15,386 @@
-   *  - STO == true:    unhold Controller + recover Drives
+   * @brief Stop the state machine.
+   *
+   * @note This is only used in tests.
```
No this is also done in the destructor

martiniil replied:
Yes, I reformulated it.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/sto_state_machine.h:
```diff
@@ -0,0 +1,342 @@
+ * @brief An AsyncStoTask is represented by a task execution and a completion signalling.
+ *
+ * The separation of task execution and the completion signalling allows the task execution to be done asynchronously.
+ * Both functions have the signature @code void() @endcode.
```
I can see that ;-)

martiniil replied:
Of course you can ;-)

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -15,94 +15,386 @@
+  //! Flag indicating if the worker-thread should terminate
+  std::atomic_bool terminate_{false};
+
+  //! Mutex for protecting access to the state machine
```
We should more carefully explain where this mutex **needs** to be used

martiniil replied:
Done

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h:
```diff
@@ -15,94 +15,386 @@
 }
-#endif // ADAPTER_STO_H
+template <class T>
+void AdapterStoTemplated<T>::workerThreadFun()
```
Could be directly a lambda on construction

martiniil replied:
I know. But I would prefer code readability over micro-optimization in this case.

agutenkunst commented on prbt_hardware_support/include/prbt_hardware_support/sto_state_machine.h:
```diff
@@ -0,0 +1,342 @@
+  /**
+   * @brief Signal completion of the task execution.
+   */
+  void signalCompletion()
```
"Signal"?

martiniil replied:
Why not?

agutenkunst committed:
Add comment about used pattern

agutenkunst committed:
Add comment for better threading understanding

agutenkunst committed:
Add ref between specs

agutenkunst commented on prbt_hardware_support/spec.dox:
```diff
@@ -27,28 +27,35 @@
  * @definereq{Stop1_Trigger}
  *  - A Stop 1 is triggered, if at least one of the following cases is true:
- *    - the state of the STO changes from TRUE -> FALSE,
+ *    - the state of the STO changes from TRUE -> FALSE when the drives are not recovering,
```
Added link to relevant Spec, @martiniil please check

martiniil replied:
Thanks! Looks good.

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_modbus_adapter_sto.cpp:
```diff
@@ -208,50 +218,59 @@ TEST_F(ModbusAdapterStoTest, testSetupNoDisableService)
-  auto t = asyncConstructor();
-  ros::Duration(5.5).sleep(); // slightly longer than WAIT_FOR_SERVICE_TIMEOUT_S
+  auto adapter_thread = asyncConstructor();
+  ros::Duration(1.0).sleep();
```
Why suddenly less sleep?

martiniil replied:
Due to `WAIT_FOR_SERVICE_TIMEOUT_S=5.0` this test checked that there is no timeout. However the constant no longer exists. So the duration is arbitrarily chosen.

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_modbus_adapter_sto.cpp:
```diff
@@ -318,46 +340,125 @@ TEST_F(ModbusAdapterStoTest, testRemoveUnholdService)
+    }
+    });
+
+  BARRIER("unhold_callback");
```
If I understand this correctly we support unholding if we know there is no Hold Service?
This is also not implied by the Spec Item "Controller_service_hold_optional"

martiniil replied:
Good point. But there is also no contrary statement in the Spec.

Needs to be discussed. Probably the Spec has to be refined. I see a few options:
- check `hold_srv_client_.exists()` before calling unhold. But what if the preceding hold failed for another reason?
- deny all attempts to unhold once a controller service failed. 
- make all services required? That is always shut down when something unexpected happens?

What do you think?

martiniil replied:
Or just print out a warning stating that a Stop1 might no longer be feasible? And that the hardware might be damaged by further stops?
After all we could never guarantee a Stop1. So we shouldn't make a big fuss about it, right?

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_modbus_adapter_sto.cpp:
```diff
@@ -318,46 +340,125 @@ TEST_F(ModbusAdapterStoTest, testRemoveUnholdService)
+    }
+    });
+
+  BARRIER("unhold_callback");
```
Not specified by "Controller_service_is_executing_optional" and needs to be discussed! Does not seem like a good idea to me

martiniil replied:
See above.

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_modbus_adapter_sto.cpp:
```diff
@@ -524,74 +674,53 @@ TEST_F(ModbusAdapterStoTest, testNoSto)
-    updateSto(sto);
-  }
-};
+  ModbusMsgStoWrapperException* exception = new ModbusMsgStoWrapperException("test");
```
Please use shared pointer. Maybe seems silly a first but a codebase without a delete / new would be nice to have.

martiniil replied:
Done.

agutenkunst commented on prbt_hardware_support/test/integrationtests/integrationtest_stop1.cpp:
```diff
@@ -231,12 +234,13 @@ TEST_F(Stop1IntegrationTest, testServiceCallbacks)
 
     // Call from Disconnect
     EXPECT_CALL(manipulator, holdCb(_,_)).Times(1);
+    EXPECT_CALL(manipulator, isExecutingCb(_,_)).Times(AtLeast(1));
```
Why don't we know the exact time this happens?

martiniil replied:
Well, because the service is called repeatedly. But I guess it was not designed for that and potentially leads to instabilities ...

After thinking it through I see the problem caused by the controller implementation. In my opinion the controller should either
- provide means for waiting until the hold position is reached
- wait until hold position is reached before returning the hold service. PR: #212

In order to make progress I will proactively use #212 to resolve the issues with is_executing in this context.

agutenkunst commented on prbt_hardware_support/test/unittests/unittest_adapter_sto.cpp:
```diff
@@ -0,0 +1,827 @@
+ *     let recover and unhold services return success
+ *  2. Call updateSto(false)),
+ *     let hold and halt services return success, let is_executing service return "executing" once and then "not executing"
+ *  3. Call updateSto(true)) repeatedly,
```
Please explain a bit more the reason behind the spaming...

agutenkunst replied:
Seems to me critical to use parallel threads in the test. Would be much more deterministic to run all the tests in parallel.

martiniil replied:
The reason is that an sto update can happen anytime during a service call. But it should have no effect, if the sto didn't change.

Concerning your second comment: I do not see a need to add more parallel threads in the tests. So KISS

agutenkunst replied:
Ok. First there was i missunderstanding. I am critical of having two test-threads and not just one (KISS)

agutenkunst commented on prbt_hardware_support/test/unittests/unittest_adapter_sto.cpp:
```diff
@@ -0,0 +1,827 @@
+
+#define EXPECT_IS_EXECUTING                                  \
+  EXPECT_CALL(mock_factory_, call_named(IS_EXECUTING_SERVICE, _)) \
+      .WillOnce(Invoke(isExecutingInvokeAction(true))) \
```
Why return true and then false? Better split into two EXPECT?

martiniil replied:
WIll be outdated if the controller change gets accepted.

agutenkunst commented on prbt_hardware_support/test/unittests/unittest_adapter_sto.cpp:
```diff
@@ -0,0 +1,827 @@
+/*
```
Various TestCases are printing
```
GMOCK WARNING:
Uninteresting mock function call - returning default value.
    Function call: handle_invalid_named(@0x7fded4f9c760 "manipulator_joint_trajectory_controller/is_executing")
          Returns: false
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#knowing-when-to-expect for details.
```
Please fix

agutenkunst replied:
Doing `stress --cpu 8 --io 4 --vm 2` and `for i in {0..100}; do catkin_make run_tests_prbt_hardware_support_gtest_unittest_adapter_sto; done`

results sometimes in stuck testcase of either one of
- AdapterStoTest.testSpamEnablePlusStop
- AdapterStoTest.testIsExecutingDelayedFail
- AdapterStoTest.testIsExecutingFail


Sadly I cannot reproduce with Level DEBUG showing the state transition, but this indicates some timing issue where I would suspect the spaming to be involved...

agutenkunst replied:
I suggest dropping the spam. I figure you needed it to achieve full coverage?

agutenkunst replied:
Simply changing the debug level (log4j.logger.ros.prbt_hardware_support.STOStateMachine=DEBUG) has a huge impact on the test result. This unittest should be **re-written in a purely deterministic way**.

martiniil replied:
The problems seem to be resolved after removing the is_executing calls. I would not re-write the tests since I really focused on being deterministic.
However if you can figure a concrete part of the test that is not deterministic, please show me.

agutenkunst replied:
Having tests fail here and there suggested some kind of non deterministic behaviour. I will have a closer look at it.

agutenkunst replied:
At 34aa704b5e66fc2c25dd1c033cbeea8cc7f6ca93

I ran `for i in {0..100}; do catkin_make run_tests_prbt_hardware_support_gtest_unittest_adapter_sto; done`
an immediatly got stuck with:
- AdapterStoTest.testSpamEnablePlusStop

agutenkunst replied:
Also

- AdapterStoTest.testUnholdFail get stuck

agutenkunst committed:
Better statemachine output

agutenkunst reviewed (changes requested):
I fix some minor things myself, see commit... Please take care about the instable tests

agutenkunst committed:
Merge master and fixes

agutenkunst committed:
Add missing EXPECTATION

agutenkunst committed:
Better debug output of state machine

agutenkunst committed:
Rename testcase

martiniil committed:
Minor changes from review

martiniil committed:
Move to hold position during hold service

(cherry picked from commit a8a86eb1d02f3f6d5a77217a1f151a38fedcb9b2)

martiniil committed:
Remove repeated is_executing service calls from adapter_sto

The usage of is_executing was not ideal in this context and a potential cause of instabilities.
See #212

martiniil committed:
Extend error message for failing controller services

agutenkunst committed:
Add missing lock for terminate_ signal

agutenkunst committed:
Add State EnableRequestDuringStop to StoAdapter

agutenkunst committed:
Fix test

agutenkunst committed:
Better BARRIER function

agutenkunst committed:
Extend unittest_adapter_stp

agutenkunst committed:
Tidy up the output

agutenkunst committed:
Minor test fix

agutenkunst commented:
Just a short update before closing. Now 
```
[DEBUG] [1566306953.779437048]: updateSto(true)
[DEBUG] [1566306953.779504680]: Event: sto_updated - Leaving: RobotInactive
[DEBUG] [1566306953.779529426]: Event: sto_updated - Action: recover_start
[DEBUG] [1566306953.779558091]: Event: sto_updated - Entering: Enabling
[DEBUG] [1566306953.779640250]: Calling Recover (Service: /prbt/driver/recover)
[ INFO] [1566306953.781137748]: Recovering XXX
EMCY: 83#0000000000000000
[DEBUG] [1566306953.875349591]: updateSto(false)
[DEBUG] [1566306953.875440075]: Event: sto_updated - Leaving: Enabling
[DEBUG] [1566306953.875489740]: Event: sto_updated - Entering: StopRequestedDuringEnable
EMCY: 84#0000000000000000
EMCY: 85#0000000000000000
EMCY: 86#0000000000000000
EMCY: 87#0000000000000000
EMCY: 88#0000000000000000
[DEBUG] [1566306953.889808154]: Event: recover_done - Leaving: StopRequestedDuringEnable
[DEBUG] [1566306953.889886441]: Event: recover_done - Action: halt_start
[DEBUG] [1566306953.889918936]: Event: recover_done - Entering: Stopping
[DEBUG] [1566306953.889961051]: Calling Halt on driver (Service: /prbt/driver/halt)
[ INFO] [1566306953.891502354]: Halting down XXX
[DEBUG] [1566306953.891690906]: Event: halt_done - Leaving: Stopping
[DEBUG] [1566306953.891762494]: Event: halt_done - Entering: RobotInactive
```
which are **11ms delay** between recover and halt

agutenkunst committed:
Merge branch 'melodic-devel' into fix/delayed_sto
